File: 백준/Gold/17404. RGB거리 2/RGB거리 2.py
import sys

def solution():
  N = int(sys.stdin.readline())
  graph = []
  for _ in range(N):
    graph.append(list(map(int, sys.stdin.readline().split())))
  min_num = 1001 * N
  for i in range(3):
    dp = [[1001 * N, 1001 * N, 1001 * N] for _ in range(N)]
    dp[0][i] = graph[0][i]
    for j in range(1, N):
      dp[j][0] = graph[j][0] + min(dp[j-1][1], dp[j-1][2])
      dp[j][1] = graph[j][1] + min(dp[j-1][0], dp[j-1][2])
      dp[j][2] = graph[j][2] + min(dp[j-1][0], dp[j-1][1])
    for k in range(3):
      if k != i:
        min_num = min(min_num, dp[N-1][k])
  print(min_num)

solution()

File: 백준/Gold/17404. RGB거리 2/RGB거리 2.js
let fs = require('fs');
let inputs = fs.readFileSync('/dev/stdin').toString().split('\n');
const N = Number(inputs[0]);

const answer = [];

for (let j = 0; j < 3; j++) {
    const dp = Array.from({length: N}, () => Array.from({length: 3}, () => 0));
    dp[0][0] = Infinity;
    dp[0][1] = Infinity;
    dp[0][2] = Infinity;
    for (let i = 0; i < N; i++) {
        const line = inputs[i + 1].split(' ').map(Number);
        if (i === 0) {
            dp[i][j] = line[j];
        } else {
            dp[i][0] = line[0] + Math.min(dp[i - 1][1], dp[i - 1][2]);
            dp[i][1] = line[1] + Math.min(dp[i - 1][0], dp[i - 1][2]);
            dp[i][2] = line[2] + Math.min(dp[i - 1][0], dp[i - 1][1]);
        }
    }
    answer.push(Math.min(...dp[N - 1].filter((_, i) => i !== j)));
}
console.log(Math.min(...answer));

